subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/IsolatingIncrementalAnnotationProcessingIntegrationTest.groovy

```diff
@@ -63,6 +61,73 @@ class IsolatingIncrementalAnnotationProcessingIntegrationTest extends AbstractIn
         outputs.recompiledFiles("A", "AHelper", "AHelperResource.txt")
     }
 
+    def "generated files are recompiled when annotated file is affected by a change"() {
+        given:
+        def util = java "class Util {}"
+        java """
+            @Helper class A {
+                private Util util = new Util();
+            }
+        """
+
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        util.text = "class Util { public void foo() {} }"
+        run "compileJava"
+
+        then:
+        outputs.recompiledFiles("Util", "A", "AHelper", "AHelperResource.txt")
+    }
+
+    def "classes depending on generated files are recompiled when annotated file's ABI is affected by a change"() {
+        given:
+        def util = java "class Util {}"
+        java """
+            @Helper class A {
+                public Util util = new Util();
+            }
+        """
+        java """
+            class Dependent {
+                private AHelper helper = new AHelper();
+            }
+       """
+
+        outputs.snapshot { run "compileJava" }
+
+        when:
+        util.text = "class Util { public void foo() {} }"
+        run "compileJava"
+
+        then:
+        outputs.recompiledFiles("Util", "A", "AHelper", "AHelperResource.txt", "Dependent")
+    }
+
+    def "classes depending on generated files are not recompiled when annotated file's implementation is affected by a change"() {
```

Comment:
My assumption here is that a change to a private implementation detail won't realistically affect the ABI of any generated types (since these generated types have no way of accessing these internals in the source class).

Reply:
The processor can't access these details unless it already violates the processing contract and uses Javac-internal classes. We already say that incremental processors aren't allowed to do so.

Reply:
👍